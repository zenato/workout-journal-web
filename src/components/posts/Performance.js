import _ from 'lodash'
import React from 'react'
import PropTypes from 'prop-types'
import { Field } from 'redux-form'
import styled from 'styled-components'
import { FormGroup, Input, Button } from 'components/form'

const Select = styled.select`
  display: block;
  background-color: #fff;
  color: #4a4a4a;
  border: solid #d4d4d4 1px;
  padding: 10px;
  width: 100%;
`
Select.displayName = 'Select'

const renderEvent = ({ events }) => ({ input, meta: { touched, error } }) => {
  const onChange = e => {
    e.preventDefault()
    const { id } = _.find(events, { id: e.target.value }) || {}
    input.onChange(id ? { id } : null)
  }
  return (
    <div>
      <Select {...input} value={input.value.id} onChange={onChange}>
        <option>Select event</option>
        {events.map(e => (
          <option key={e.id} value={e.id}>
            {e.name}
          </option>
        ))}
      </Select>
      {touched && error && <span className="error">{error}</span>}
    </div>
  )
}

const Performance = ({ events, name, onDelete, values }) => {
  const total = _.sum(
    _.range(1, 6).map(i => {
      const value = _.toNumber(values[`set${i}`])
      return isNaN(value) ? 0 : value
    }),
  )
  const volume = total * (isNaN(values.value) ? 0 : _.toNumber(values.value))
  const selectedEvent = (values.event ? _.find(events, { id: values.event.id }) : null) || {}
  return (
    <div>
      <FormGroup label="Event">
        <Field name={`${name}.event`} component={renderEvent({ events })} />
      </FormGroup>
      <FormGroup label="Value">
        <Field name={`${name}.value`} type="text" component={Input} />
      </FormGroup>
      <FormGroup label="Unit">{selectedEvent.unit}</FormGroup>
      <FormGroup label="Set1">
        <Field name={`${name}.set1`} type="text" component={Input} />
      </FormGroup>
      <FormGroup label="Set2">
        <Field name={`${name}.set2`} type="text" component={Input} />
      </FormGroup>
      <FormGroup label="Set3">
        <Field name={`${name}.set3`} type="text" component={Input} />
      </FormGroup>
      <FormGroup label="Set4">
        <Field name={`${name}.set4`} type="text" component={Input} />
      </FormGroup>
      <FormGroup label="Set5">
        <Field name={`${name}.set5`} type="text" component={Input} />
      </FormGroup>
      <FormGroup label="Tot. / Vol.">
        {total} / {volume}
      </FormGroup>
      <FormGroup>
        <Button value="Del" onClick={onDelete} />
      </FormGroup>
    </div>
  )
}

Performance.propTypes = {
  name: PropTypes.string.isRequired,
  values: PropTypes.object.isRequired,
  events: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      last_performace: PropTypes.shape({
        value: PropTypes.number.isRequired,
      }),
    }),
  ).isRequired,
}

export default Performance
