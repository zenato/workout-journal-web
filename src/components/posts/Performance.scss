@import 'common';
@import '../form/Input.scss';
