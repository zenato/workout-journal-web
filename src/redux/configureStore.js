import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import penderMiddleware from 'redux-pender';
import modules from './modules';

const isDevelopment = process.env.NODE_ENV === 'development';

const composeEnhancers = isDevelopment
  ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
  : compose;

const configureStore = initialState => {
  const store = createStore(
    modules,
    initialState,
    composeEnhancers(applyMiddleware(thunk, penderMiddleware())),
  );

  if (module.hot) {
    module.hot.accept('./modules', () => {
      const nextRootReducer = require('./modules').default;
      store.replaceReducer(nextRootReducer);
    });
  }

  return store;
};

export default configureStore;
