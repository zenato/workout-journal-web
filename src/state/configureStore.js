import { createStore, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import thunk from 'redux-thunk'
import penderMiddleware from 'redux-pender'
import reducers from './reducers'
import rootSaga from './sagas'

const isDevelopment = process.env.NODE_ENV === 'development'

const composeEnhancers = isDevelopment
  ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
  : compose

const configureStore = initialState => {
  const sagaMiddleware = createSagaMiddleware()
  const store = createStore(
    reducers,
    initialState,
    composeEnhancers(applyMiddleware(sagaMiddleware, thunk, penderMiddleware())),
  )

  if (module.hot) {
    module.hot.accept('./reducers', () => {
      const nextRootReducer = require('./reducers').default
      store.replaceReducer(nextRootReducer)
    })
  }

  sagaMiddleware.run(rootSaga);
  return store
}

export default configureStore
