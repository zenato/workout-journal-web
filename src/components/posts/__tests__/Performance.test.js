import React from 'react';
import { Provider } from 'react-redux';
import { reduxForm } from 'redux-form'
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux/configureStore';
import Performance from '../Performance';

configure({ adapter: new Adapter() });

const store = configureStore();

const Test = (props) => (
  <Provider store={store} {...props} />
);

const Form = (props) => (<form {...props} />);
const MockForm = reduxForm({ form: 'eventForm' })(Form);

const events = [
  {
    id: '1',
    name: 'test',
    value: 10,
    unit: 'g',
    remark: 'test',
  },
];

const values = {
  id: 1,
  name: 'test',
  value: 10,
  unit: 'g',
  remark: 'test',
};

it('renders without errors', () => {
  shallow(
    <Test>
      <MockForm>
        <Performance
          key={0}
          name={'performance[0]'}
          events={events}
          values={values}
          onDelete={() => {}}
        />
      </MockForm>MockForm>
    </Test>
  );
});

it('simulate delete click', () => {
  const onDelete = jest.fn();

  const component = mount(
    <Test>
      <MockForm>
        <Performance
          key={0}
          name={'performance[0]'}
          events={events}
          values={values}
          onDelete={onDelete}
        />
      </MockForm>
    </Test>
  );

  component.find('input[type="button"]').simulate('click', { preventDefault () {} });

  expect(onDelete).toBeCalled();
});
