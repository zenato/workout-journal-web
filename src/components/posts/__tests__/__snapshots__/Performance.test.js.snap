// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance should render without errors. 1`] = `
<div>
  <FormGroup
    label="Event"
  >
    <Field
      component={[Function]}
      name="performance[0].event"
    />
  </FormGroup>
  <FormGroup
    label="Value"
  >
    <Field
      component={[Function]}
      name="performance[0].value"
      type="text"
    />
  </FormGroup>
  <FormGroup
    label="Unit"
  >
    g
  </FormGroup>
  <FormGroup
    label="Set1"
  >
    <Field
      component={[Function]}
      name="performance[0].set1"
      type="text"
    />
  </FormGroup>
  <FormGroup
    label="Set2"
  >
    <Field
      component={[Function]}
      name="performance[0].set2"
      type="text"
    />
  </FormGroup>
  <FormGroup
    label="Set3"
  >
    <Field
      component={[Function]}
      name="performance[0].set3"
      type="text"
    />
  </FormGroup>
  <FormGroup
    label="Set4"
  >
    <Field
      component={[Function]}
      name="performance[0].set4"
      type="text"
    />
  </FormGroup>
  <FormGroup
    label="Set5"
  >
    <Field
      component={[Function]}
      name="performance[0].set5"
      type="text"
    />
  </FormGroup>
  <FormGroup
    label="Tot. / Vol."
  >
    0
     / 
    0
  </FormGroup>
  <FormGroup>
    <Button
      onClick={[Function]}
      value="Del"
    />
  </FormGroup>
</div>
`;
